[flang] Register and lower SECNDS (stubbed implementation)

mlir-maiden opened the pull request:
This patch registers and lowers the GNU extension intrinsic `SECNDS` in Flang.

- Registration and lowering are wired through the intrinsic infrastructure.
- genSecnds() currently emits a TODO fatal error, marking it as unimplemented.
- Includes an XFAIL test to exercise the lowering path and reflect current status.

Fixes https://github.com/llvm/llvm-project/issues/58728

---

CC @eugeneepshteyn @klausler

mlir-maiden committed:
[flang] Register and lower SECNDS (stubbed implementation)

eugeneepshteyn commented on flang/lib/Optimizer/Builder/IntrinsicCall.cpp:
```diff
@@ -7815,6 +7819,14 @@ IntrinsicLibrary::genScan(mlir::Type resultType,
 }
 
+// SECNDS
+// Lowering is registered, runtime not yet implemented
```
Hello, thank you for the patch!

I happen to know how to add `secnds()` to the runtime part, so I can quickly do that part.

`genSecnds()` though, is not part of runtime, it's part of flang compiler that generates MLIR/FIR/HLFIR code to *set up* the runtime call.  Would you like to try to do that? That's probably going to be similar (but easier) to `genEtime()`, where the input arg is always one REAL.

eugeneepshteyn commented:
Also, could you please document this new intrinsic in https://github.com/llvm/llvm-project/blob/main/flang/docs/Intrinsics.md ?

klausler commented on flang/lib/Evaluate/intrinsics.cpp:
```diff
@@ -921,6 +921,11 @@ static const IntrinsicInterface genericIntrinsicFunction[]{
             DefaultingKIND},
         KINDInt},
+    {"secnds",
+        {{"x", TypePattern{RealType, KindCode::exactKind, 4}, Rank::scalar,
```
Are you sure that the reference argument is always `REAL(4)`, or is it the default `REAL` kind?  Are the semantics in GNU Fortran affected when the default `REAL` kind is changed on the command line?

`Optionality::required` and `Intent::In` are the default settings, so you don't need to specify them again here.

klausler commented on flang/lib/Evaluate/intrinsics.cpp:
```diff
@@ -921,6 +921,11 @@ static const IntrinsicInterface genericIntrinsicFunction[]{
+        {{"x", TypePattern{RealType, KindCode::exactKind, 4}, Rank::scalar,
+            Optionality::required, common::Intent::In}},
+        TypePattern{RealType, KindCode::exactKind, 4}, Rank::scalar,
+        IntrinsicClass::impureSubroutine},
```
I think that `SECNDS` is a function.

eugeneepshteyn commented:
Runtime PR: https://github.com/llvm/llvm-project/pull/152021

mlir-maiden committed:
[flang] Register and lower SECNDS intrinsic (stubbed implementation)

- Updated SECNDS handler to pass `refTime` asAddr instead of asValue, matching the backend runtime contract (`float*`).
- Addressed review feedback in the lowering hook:
  * Built the argument list using fir::runtime::createArguments
  * Replaced previous call emission with fir::CallOp::create(...)
- Updated the regression test for SECNDS lowering to check argument passing, call signature, and return type.
- Added documentation for SECNDS to Intrinsics.md, including description, usage, and example.

klausler commented:
> @mlir-maiden , I submitted runtime implementation for `secnds()` upstream. You can now test end-to-end with trying to make the actual function call.
> 
> In my implementation I defined runtime wrapper function for `secnds()` as follows:
> 
> ```
> float RTNAME(Secnds)(float *refTime, const char *sourceFile, int line) {
>   Terminator terminator{sourceFile, line};
>   RUNTIME_CHECK(terminator, refTime != nullptr);
>   return FORTRAN_PROCEDURE_NAME(secnds)(refTime);
> }
> ```
> 
> Feel free to modify, if that doesn't work for you.
> 
> Peter also asked me to make the core implementation of `secnds` runtime into a template, so that it could be instantiated with `double` in case we want to also add `DSECNDS()`. (If you end up implementing `DSECNDS()`, I recommend you do that as a separate PR.)

It's `SECNDSD` and it might be just a PGI extension (also in "classic" flang).

eugeneepshteyn commented on flang-rt/lib/runtime/time-intrinsic.cpp:
```diff
@@ -576,5 +576,8 @@ void RTNAME(Etime)(const Descriptor *values, const Descriptor *time,
 }
 
+// Stub for SECNDS intrinsic; not yet implemented
+float RTNAME(Secnds)(float refTime) { TODO("not yet implemented: SECNDS"); }
```
Note that in Fortran the args are generally passed by reference, hence in my PR this is `float *refTime`.

klausler replied:
This should be the case for routines called by Fortran, but it need not be so for runtime library entry points called only by lowered code, like this.

eugeneepshteyn commented on flang/include/flang/Runtime/time-intrinsic.h:
```diff
@@ -46,6 +46,8 @@ void RTNAME(DateAndTime)(char *date, std::size_t dateChars, char *time,
 void RTNAME(Etime)(const Descriptor *values, const Descriptor *time,
     const char *sourceFile, int line);
 
+float RTNAME(Secnds)(float refTime);
```
`float *refTime`

eugeneepshteyn commented on flang/lib/Optimizer/Builder/Runtime/Intrinsics.cpp:
```diff
@@ -276,6 +276,15 @@ void fir::runtime::genRename(fir::FirOpBuilder &builder, mlir::Location loc,
+                                    mlir::Location loc, mlir::Value refTime) {
+  auto runtimeFunc =
+      fir::runtime::getRuntimeFunc<mkRTKey(Secnds)>(loc, builder);
+  return builder
+      .create<fir::CallOp>(loc, runtimeFunc, mlir::ValueRange{refTime})
+      .getResult(0);
```
Wouldn't it be something like `return fir::CallOp::create(builder, loc, runtimeFunc, args).getResult(0);` ?  And need to `fir::runtime::createArguments()` to create `args` for `refTime`.

eugeneepshteyn commented on flang/lib/Evaluate/intrinsics.cpp:
```diff
@@ -2687,7 +2691,7 @@ std::optional<SpecificCall> IntrinsicInterface::Match(
   ActualArguments rearranged(dummies);
   for (std::size_t j{0}; j < dummies; ++j) {
-    if (ActualArgument *arg{actualForDummy[j]}) {
+    if (ActualArgument * arg{actualForDummy[j]}) {
```
This space between `*` and `arg` seems to be wrong. Did you run your changes through `clang-format`?

eugeneepshteyn commented on flang/lib/Evaluate/intrinsics.cpp:
```diff
@@ -3588,7 +3592,7 @@ std::optional<SpecificCall> IntrinsicProcTable::Implementation::Probe(
   auto specificRange{specificFuncs_.equal_range(call.name)};
   for (auto specIter{specificRange.first}; specIter != specificRange.second;
-       ++specIter) {
+      ++specIter) {
```
Make sure that `clang-format` is happy with it.

eugeneepshteyn commented on flang/test/Lower/Intrinsics/secnds.f90:
```diff
@@ -0,0 +1,12 @@
+  x = secnds(1.0)
+end program
+
+! CHECK: not yet implemented: SECNDS
```
Now you should be able to check for actual MLIR.

mlir-maiden committed:
Fix pointer spacing and indentation

mlir-maiden reviewed (commented):
So nice to e-meet you, Peter and Eugene. What an introduction - you won’t believe what happened! I have been quickly replying to your review comments as they came in over the past week (and was so proud of it!), only to find out today that, due to a GitHub review workflow quirk, they were never actually sent. I’m still getting the hang of GitHub, so my apologies. 
I’m submitting them now so they’re finally visible. I do want to work on extending DSECNDS, and thank you, Eugene, for the tip+template. It’s very helpful. Glad to finally be “online” with you both now.
@eugeneepshteyn  @klausler

mlir-maiden commented:
@klausler @eugeneepshteyn 
Hello team, just a gentle nudge on this. I believe the outstanding comments have been addressed, but please let me know if there’s anything further needed on my side. Thank you for your time and review!

eugeneepshteyn commented:
> ...I believe the outstanding comments have been addressed, but please let me know if there’s anything further needed on my side. 

I checked with Peter and he's ok with this going in. I'm going to check again today to ensure this still works with the latest upstream, and will then approve this PR. 

I don't think you have privileges to merge this in, so I'll do the merge myself.

eugeneepshteyn committed:
Merge branch 'main' into secnds-lowering
